### ISSUE TITLE
orphaned bridge & virt interfaces

### ISSUE THREAD

**Issue by anapsix** [docker/docker#12239]
It appears Docker does not cleanup bridges & virtual interfaces upon container shutdown.
I came across this while setting up Deis, which creates a lot of containers, especially when some of them are having trouble starting (due to misconfiguration, etc).

Issue created per @thaJeztah suggestion. Problem was mentioned at deis/deis#3336 and is referred to in #1320 by @ianblenke, @Blystad.

Despite using newer version of CoreOS (633.1.0 beta channel), I'm having same issue with interfaces maxing out:
```
anapsix@core1 ~ $ brctl show | wc -l
1024

anapsix@core1 ~ $ ip -o link sh | wc -l
18191

anapsix@core1 ~ $ docker info
Containers: 8
Images: 117
Storage Driver: overlay
 Backing Filesystem: extfs
Execution Driver: native-0.2
Kernel Version: 3.19.0
Operating System: CoreOS 633.1.0
CPUs: 24
Total Memory: 62.91 GiB
Name: core1.internal
ID: EGWF:6EOB:VPOB:EY5Z:SBLM:FGHL:HG5Y:XA5K:OGAW:EZ6U:OWH2:WOJZ

anapsix@core1 ~ $ docker version
Client version: 1.5.0
Client API version: 1.17
Go version (client): go1.3.3
Git commit (client): a8a31ef-dirty
OS/Arch (client): linux/amd64
Server version: 1.5.0
Server API version: 1.17
Go version (server): go1.3.3
Git commit (server): a8a31ef-dirty
```

**Comment by rohansingh** [comment]
Seeing this too on many hosts at Spotify when upgrading to 1.6 (from 1.3). @anapsix, when this happens, can you check with `ps -aux` and see if there are orphaned container processes left around?

When we see this, generally what's happening is that a container process is not killed by Docker, even though Docker thinks the container has stopped. The process is still around, it still has a network namespace, virtual ethernet device, and a configured IP. Leads to fun problems with traffic going to "dead" containers as well.

**Comment by thaJeztah** [comment]
@rohansingh Out of interest, did Spotify switch directly from 1.3 to 1.6, or have releases between that also been used/tested (to narrow down if this problem was introduced in 1.6 or an earlier release).

Networking will get a major overhaul in Docker 1.7 with the initial integration of Libnetwork (https://github.com/docker/libnetwork), but I doubt that will help if containers are not stopped/killed properly.

Would you be able to provide some more information on the setup you're using, starting with the output of;

- `uname -a`
- `docker info`
- `docker version` (*not* `docker --version`, which only reports the version)

/cc @LK4D4 @mrjana @crosbymichael

**Comment by rohansingh** [comment]
@thaJeztah Off the top of my head though, we upgraded from 1.3.2 directly to 1.6.0. On some hosts, we run a service container and a diagnostic container.

After the upgrade, the diagnostic container started getting requests that were bound for the service container on some percentage of the hosts (~40 out of ~hundreds). These were requests being sent to a port mapped to the service container as reported by `docker ps`. Our repro showed this would happen with requests from other hosts and on localhost (i.e., `docker-proxy` was affected so it clearly wasn't just an iptables issue).

We reproduced in a test environment by running two containers and a script that does something like:

```bash
while true; do
  apt-get install lxc-docker-1.6.0
  sleep 60
  apt-get install lxc-docker-1.3.2
  sleep 60
done
```

The contrived containers we're running during this process are using some Alpine image (`onescience/alpine`) and running something like this:

1. `nc -lk -p 5700 -e /bin/sh -c 'container A'`
2. `nc -lk -p 5701 -e /bin/sh -c 'container B'`

The two containers are started by Helios whenever it detects they are missing from `docker ps`, though this should also be reproducible by putting some `docker run` statements in that script above.

We run something like `while true; do nc -v localhost 5701 2>&1 | grep "container A"; done` on the host as well. If anything is printed to stdout, it means we received `container A` on port 5701, which should never happen. That said, after a few minutes of running the script above, it does happen pretty consistently.

Upon reproducing, `ps` shows many old container processes (i.e, `nc -lk ...`) still running, though `docker ps` shows Docker is only aware of the two newest containers.

`ip netns exec ... ifconfig` shows some of the old "container A" processes have devices configured with the same IP as the newest instance of "container B". Not surprising, since the order we start "container A" and "container B" in is somewhat random, so the IP's they're assigned flip-flops after a Docker restart.

`ip netns exec ... tcpflow -c -i eth0` shows that when this happens, old instances of "container A" are indeed receiving the traffic bound for the newest "container B".

Meanwhile, `ifconfig` on the host shows plenty of lingering veth devices, which makes sense based on the above. That common symptom is how @rculbertson found this open issue.

---

This started on Wednesday and we just managed to repro and figure out the details above at the end of the day on Friday. On Monday, I'm planning on checking if this is only an issue with 1.3.2 (fingers crossed).

That is to say, if we can repro by just repeatedly restarting Docker 1.3.2 but can't repro by repeatedly restarting Docker 1.6.0, then I'd be comfortable saying it's a pre-1.6 issue. Alternatively, if we can't repro by switching between `lxc-docker-1.6.0` and `lxc-docker-1.6.2`, or if we can somehow prove that the orphans were all spawned and left behind by 1.3.2, I'd also be comfortable saying it's a pre-1.6 issue.

We're using Ubuntu 14.04, stock kernel, vanilla Docker packages from get.docker.com/ubuntu. I'll gather the `uname` and version details and reply back.

**Comment by rohansingh** [comment]
Details from a experiment box we've successfully repro'd on, which I guess has 1.3.2 on it now:

```
rohan@lon3-heliosagent-a1:~$ uname -a
Linux lon3-heliosagent-a1.lon3.spotify.net 3.13.0-19-generic #40-Ubuntu SMP Mon Mar 24 02:36:06 UTC 2014 x86_64 x86_64 x86_64 GNU/Linux

rohan@lon3-heliosagent-a1:~$ docker info
Containers: 2
Images: 9
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Dirs: 57
Execution Driver: native-0.2
Kernel Version: 3.13.0-19-generic
Operating System: Ubuntu Trusty Tahr (development branch)
Debug mode (server): true
Debug mode (client): false
Fds: 25
Goroutines: 23
EventsListeners: 0
Init Path: /usr/bin/docker
WARNING: No swap limit support

rohan@lon3-heliosagent-a1:~$ docker version
Client version: 1.3.2
Client API version: 1.15
Go version (client): go1.3.3
Git commit (client): 39fa2fa
OS/Arch (client): linux/amd64
Server version: 1.3.2
Server API version: 1.15
Go version (server): go1.3.3
Git commit (server): 39fa2fa

rohan@lon3-heliosagent-a1:~$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu Trusty Tahr (development branch)
Release:	14.04
Codename:	trusty
```

Details from a production box that actually had this issue, which is in its post-upgrade state:

```
rohan@ash2-presenceview-a2:~$ uname -a
Linux ash2-presenceview-a2.ash2.spotify.net 3.13.0-34-generic #60-Ubuntu SMP Wed Aug 13 15:45:27 UTC 2014 x86_64 x86_64 x86_64 GNU/Linux

rohan@ash2-presenceview-a2:~$ docker info
Containers: 1
Images: 22
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 24
 Dirperm1 Supported: false
Execution Driver: native-0.2
Kernel Version: 3.13.0-34-generic
Operating System: Ubuntu 14.04 LTS
CPUs: 24
Total Memory: 31.25 GiB
Name: ash2-presenceview-a2.ash2.spotify.net
ID: RCJY:65F3:X4AG:P7RU:2EUG:VGIT:IUT7:FMAZ:Q6LD:BNU3:O54F:JNYP
Debug mode (server): true
Debug mode (client): false
Fds: 22
Goroutines: 30
System Time: Sat May 16 17:20:18 UTC 2015
EventsListeners: 0
Init SHA1: 9145575052383dbf64cede3bac278606472e027c
Init Path: /usr/bin/docker
Docker Root Dir: /var/lib/docker
WARNING: No swap limit support

rohan@ash2-presenceview-a2:~$ docker version
Client version: 1.6.0
Client API version: 1.18
Go version (client): go1.4.2
Git commit (client): 4749651
OS/Arch (client): linux/amd64
Server version: 1.6.0
Server API version: 1.18
Go version (server): go1.4.2
Git commit (server): 4749651
OS/Arch (server): linux/amd64

rohan@ash2-presenceview-a2:~$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 14.04 LTS
Release:	14.04
Codename:	trusty
```

**Comment by thaJeztah** [comment]
@rohansingh :heart: thanks so much for the detailed information. Debugging these is not really my expertise, but I'm trying to collect the relevant information, so that the people that *are* able to check can do so.

I'll attempt to get the right people in this discussion to get this moving (or see if improvements were already made in this area)

**Comment by anapsix** [comment]
@rohansingh thanks for a a backup and most thorough repro.. and a name drop.. don't know if it would be notices otherwise :/
I don't remember seeing lingering processes, but there had to be some.. which is complementary to the interfaces problem.
I'll try to reproduce on my side, paying attention to the process list.

**Comment by rohansingh** [comment]
Looks like the problem is with the Docker daemon shutdown process. There are multiple issues:

1. For some reason, `Daemon.shutdown` has its own process for shutting down containers rather than using `Container.Stop`. It only ever sends SIGTERM, never SIGKILL, so a container that refuses to die on SIGTERM will survive a daemon shutdown.

2. Even though `Daemon.shutdown` will wait forever for containers to stop, `Engine.Shutdown` only gives shutdown handlers 10 seconds to run. After that, the engine will shutdown and any surviving containers will still be around.

3. Even though `Engine.Shutdown` will give all handlers 15 seconds to complete (5 seconds for tasks, 10 for shutdown handlers), the Upstart script in the official `lxc-docker` packages doesn't override the default Upstart kill timeout of 5 seconds. Since Upstart will send SIGKILL to the Docker daemon just five seconds after SIGTERM, shutdown handlers may never run.

TL;DR — Shutdown of the daemon is kind of broken. I'll submit a PR to address these.

**Comment by thaJeztah** [comment]
@rohansingh I think I've seen some PRs recently on that matter, but there were some issue around what the "right" approach was, so would be wise to check existing PRs as well.

**Comment by rohansingh** [comment]
@thaJeztah Thanks, looks like #13304 should be addressing this. I'll add my comments there.

**Comment by thaJeztah** [comment]
@rohansingh Sounds like  a candidate, yes. Given that you probably have some servers to test/repro, it could be helpful :)

**Comment by aespinosa** [comment]
Hi @thaJeztah 

Which release was this fixed?  We're also seeing this in 1.7.1

**Comment by bacongobbler** [comment]
It looks like #13304 was the latest PR which was merged into v1.9.0.

```
$ git log v1.8.3 v1.9.0 --oneline | grep 13304
2cec06f Merge pull request #13304 from coolljt0725/restart_daemon_with_paused_containers
```

**Comment by aboch** [comment]
@anapsix @rohansingh @aespinosa We have not seen this issue for quite sometime.
Given the above mentioned changes were merged, can you confirm the issue is not reproducible for you as well with recent docker versions ?
Thanks.

**Comment by anapsix** [comment]
I don't always loose interest in the issues I've reported, but when I do, it's been over a year..

**Comment by anapsix** [comment]
also.. haven't encountered it since

**Comment by thaJeztah** [comment]
I'll go ahead and close this issue, but feel free to comment if you think I closed this by mistake
